Clustered backup and restore (#5833)
influxdata/docs-v2: content/influxdb3/clustered/admin/backup-restore.md

in case of emergency, we recommend enabling snapshots.

With InfluxDB Clustered snapshots enabled, each hour, InfluxDB uses the `pg_dump`
utility included with the InfluxDB Garbage Collector to export an SQL blob or
“snapshot” from the InfluxDB Catalog and store it in the object store.
The Catalog is a PostgreSQL-compatible relational database that stores metadata
for your time series data, such as schema data types, Parquet file locations, and more.

The Catalog snapshots act as recovery points for your InfluxDB cluster that
reference all Parquet files that existed in the object store at the time of the
snapshot. When a snapshot is restored to the Catalog, the Compactor
“[soft deletes](#soft-delete)” any Parquet files not listed in the snapshot.

> [!Note]
> If a Parquet file is listed as part of *any* hourly or daily snapshot,
> InfluxDB never removes the Parquet file from the object store (also known as a
> “[hard delete](#hard-delete)”).
> 
> For example–if you have Parquet files A, B, C, and D, and you restore to a
> snapshot that includes B and C, A and D are soft-deleted, but remain in object
> storage until they are no longer referenced in any Catalog snapshot.

> 

Reviewer: ```suggestion
- [Soft delete](#soft-delete)
- [Hard delete](#hard-delete)
```

> - [RPO and RTO](#rpo-and-rto)

Reviewer: ```suggestion
- [Recovery Point Objective (RPO)](#recovery-point-objective-rpo)
- [Recovery Time Objective (RTO)](#recovery-time-objective-rto)
```

> - [Things to know](#things-to-know)

Reviewer: ```suggestion
```

>   - [Data written just before a snapshot may not be present after restoring](#data-written-just-before-a-snapshot-may-not-be-present-after-restoring)

Reviewer: ```suggestion
- [Data written just before a snapshot may not be present after restoring](#data-written-just-before-a-snapshot-may-not-be-present-after-restoring)
```

- [Recommendations](#recommendations)
  - [Automate object synchronization to an external S3-compatible bucket](#automate-object-synchronization-to-an-external-s3-compatible-bucket)
  - [Enable short-term object versioning](#enable-short-term-object-versioning)
- [Configure snapshots](#configure-snapshots)
  - [Environment Variables](#environment-variables)
- [Verify snapshots](#verify-snapshots)
- [Restore to a recovery point](#restore-to-a-recovery-point)
- [Definitions](#definitions)
- [Resources](#resources)
  - [*prep_pg_dump.awk*](#preppgdumpawk)

> 

Reviewer: ```suggestion

### Soft delete

A _soft delete_ refers to when, on compaction, the Compactor sets a `deleted_at`
timestamp on the Parquet file entry in the Catalog.
The Parquet file is no
longer queryable, but remains intact in the object store. 

## Hard delete

A _hard delete_ refers to when a Parquet file is actually deleted from object
storage and no longer exists.

```

> ## RPO and RTO
> 
> With the InfluxDB Clustered snapshot strategy, you can expect the following:
> 
> - [**Recovery Point Objective (RPO)**](#recovery-point-objective-rpo):  
>   - 1 hour for hourly snapshots *(up to the configured hourly snapshot expiration)*  
>   - 1 day for daily snapshots *(up to the configured daily snapshot expiration)*  
> - [**Recovery Time Objective (RTO)**](#recovery-time-objective-rto):
>   *RTO varies depending on the size of the Catalog database, network speeds
>   between the client machine and the Catalog database, cluster load, the status
>   of your underlying hosting provider, and other factors.*

Reviewer: ```suggestion
## Recovery Point Objective (RPO)

RPO is the maximum amount of data loss (based on time) allowed after a disruptive event.
It indicates how much time can pass between data snapshots before data is considered lost if a disaster occurs.

The InfluxDB Clustered snapshot strategy RPO allows for the following maximum data loss:

  - 1 hour for hourly snapshots *(up to the configured hourly snapshot expiration)*  
  - 1 day for daily snapshots *(up to the configured daily snapshot expiration)*
 
 ## Recovery Time Objective (RTO)

RTO is the maximum amount of downtime allowed for an InfluxDB cluster after a failure.
RTO varies depending on the size of your Catalog database, network speeds
  between the client machine and the Catalog database, cluster load, the status
  of your underlying hosting provider, and other factors.
```

> ## Things to know
> 

Reviewer: ```suggestion
```

> ### Data written just before a snapshot may not be present after restoring

Reviewer: ```suggestion
## Data written just before a snapshot may not be present after restoring
```

Due to the variability of flushing data from Ingesters into Parquet files, data
written in the last few minutes before a snapshot may not be included.
This variability is typically less than 15 minutes, but is per table.
This means that one table may have data written up to the timestamp of the
snapshot, while another may not have data written in the 15 minutes prior to the
snapshot. All data written more than 15 minutes prior to a snapshot should be
present after restoring to that snapshot.

## Recommendations

### Automate object synchronization to an external S3-compatible bucket

Syncing objects to an external S3-compatible bucket ensures an up-to-date backup
in case your object store becomes unavailable. Recovery point snapshots only
back up the InfluxDB Catalog. If data referenced in a Catalog snapshot does not
exist in the object store, the recovery process does not restore the missing data.

### Enable short-term object versioning

> If your object storage provider supports it, we recommend enabling short-term
> object versioning on your object store. As objects are updated, the object store
> retains distinct versions of each update which can be used to “rollback” newly
> written or updated Parquet files to previous versions. Storing versioned objects
> does add to object storage costs, but keeping versioned objects for 1-2 days
> provides a short buffer to protect against potential errant writes or deleted objects.

Reviewer: ```suggestion
If your object storage provider supports it, consider enabling short-term
object versioning on your object store--for example, 1-2 days to protect against errant writes or deleted objects.
With object versioning enabled, as objects are updated, the object store
retains distinct versions of each update that can be used to “rollback” newly
written or updated Parquet files to previous versions.
Keep in mind, storing versioned objects
does add to object storage costs.
```

## Configure snapshots

Use the available environment variables to enable and configure hourly Catalog
snapshots in your InfluxDB cluster. Add these environment variables to the
Garbage Collector configuration in your `AppInstance` resource:

```yaml
apiVersion: kubecfg.dev/v1alpha1
kind: AppInstance
metadata:
  name: influxdb
  namespace: influxdb
spec:
  package:
    spec:
      components:
        garbage-collector:
          template:
            containers:
              iox:
                env:
                  INFLUXDB_IOX_CREATE_CATALOG_BACKUP_DATA_SNAPSHOT_FILES: true
                  INFLUXDB_IOX_DELETE_USING_CATALOG_BACKUP_DATA_SNAPSHOT_FILES: true
                  INFLUXDB_IOX_KEEP_HOURLY_CATALOG_BACKUP_FILE_LISTS: '30d'
                  INFLUXDB_IOX_KEEP_DAILY_CATALOG_BACKUP_FILE_LISTS: '90d'

>                   INFLUXDB_IOX_GC_OBJECTSTORE_CUTOFF: '14d'

Reviewer: ```suggestion
INFLUXDB_IOX_GC_OBJECTSTORE_CUTOFF: '14d'
```

```

### Environment Variables

#### INFLUXDB_IOX_CREATE_CATALOG_BACKUP_DATA_SNAPSHOT_FILES

Enable hourly Catalog snapshotting. The default is `false`. **Set to `true`**.

#### INFLUXDB_IOX_DELETE_USING_CATALOG_BACKUP_DATA_SNAPSHOT_FILES

Enable a snapshot check when deleting files to ensure the Garbage Collector does
not remove Parquet files from the object store that are associated with existing

> snapshots. The default is `false`. **Set to `true`**.

Reviewer: ```suggestion
snapshots. The default is `false`. Set to `true`:

```yaml
INFLUXDB_IOX_DELETE_USING_CATALOG_BACKUP_DATA_SNAPSHOT_FILES: true
```

```

> [!Caution]
> If set to `false` with snapshots enabled, the Garbage Collector does not check
> to see if a Parquet file is associated with existing snapshots before removing
> the Parquet file from the object store. This could result in deleting Parquet
> files needed to restore the cluster to a recovery point.

#### INFLUXDB_IOX_KEEP_HOURLY_CATALOG_BACKUP_FILE_LISTS

After this duration of time, the Garbage Collector deletes *hourly* snapshots,

> allowing the Garbage Collector to hard-delete Parquet files from the object

Reviewer: ```suggestion
allowing the Garbage Collector to [hard-delete](#hard-delete) Parquet files from the object
```

> store and the Catalog.  The default is `30d`. The recommended range is between
> `1d` and `30d`.

Reviewer: ```suggestion
store and the Catalog.  The default is `30d`. The recommended range for snapshots is between
`1d` and `30d`:

```yaml
INFLUXDB_IOX_KEEP_HOURLY_CATALOG_BACKUP_FILE_LISTS: '30d'
```

```

#### INFLUXDB_IOX_KEEP_DAILY_CATALOG_BACKUP_FILE_LISTS

After this duration of time, the Garbage Collector deletes *daily* snapshots,

> allowing the Garbage Collector to hard-delete Parquet files from the object

Reviewer: ```suggestion
allowing the Garbage Collector to [hard-delete](#hard-delete) Parquet files from the object
```

store and the Catalog. The default is `90d`. The recommended range is between
`3d` and `90d`.

> > [!Note]

Reviewer: ```suggestion
> [!Important]
```

> > Daily snapshots must expire after hourly backups.
> > `INFLUXDB_IOX_KEEP_DAILY_CATALOG_BACKUP_FILE_LISTS` must be greater than
> > `INFLUXDB_IOX_KEEP_HOURLY_CATALOG_BACKUP_FILE_LISTS`.

Reviewer: ```suggestion
> Daily snapshots must expire after hourly backups
> Make sure to set `INFLUXDB_IOX_KEEP_DAILY_CATALOG_BACKUP_FILE_LISTS` to a value greater than
> `INFLUXDB_IOX_KEEP_HOURLY_CATALOG_BACKUP_FILE_LISTS`.
```

> 

Reviewer: ```suggestion
```yaml
INFLUXDB_IOX_KEEP_DAILY_CATALOG_BACKUP_FILE_LISTS: '90d'
```

```

#### INFLUXDB_IOX_GC_OBJECTSTORE_CUTOFF

The duration of time after a Parquet file is no longer referenced in the Catalog
or included in any snapshots after which the Garbage Collector removes the
Parquet file from the Object store. The default is `14d`. The recommended range

> is between `6h` and `14d`.

Reviewer: ```suggestion
is between `6h` and `14d`:

```yaml
INFLUXDB_IOX_GC_OBJECTSTORE_CUTOFF: '14d'
```

```

## Verify snapshots

InfluxDB Clustered stores hourly and daily snapshots in the
`/catalog_backup_file_lists` path in object storage. After enabling snapshots,

> ensure that snapshots are written to the object store using clients provided by
> your object storage provider.
> 

Reviewer: ```suggestion
use clients provided by
your object storage provider to ensure that snapshots are written to the object store.

```

Hourly snapshots are taken at *approximately* the beginning of each hour

> (≈1:00, ≈2:00, ≈3:00, etc.). After enabling snapshotting, the first snapshot is
> written on or around the beginning of the next hour.

Reviewer: ```suggestion
(≈1:00, ≈2:00, ≈3:00, etc.). After you enable snapshotting, the first snapshot is
written on or around the beginning of the next hour.
```

    <!-- pytest.mark.skip -->

    ```bash
    psql CATALOG_DSN < snapshot.sql
    ```

    The exact `psql` command depends on your PostgreSQL-compatible database
    provider, their authentication requirements, and the database’s DSN.   

7.  **Restart InfluxDB Clustered components**

    1.  In your `AppInstance` resource, set `pause` to `false` or remove the `pause`:   

        ```yaml
        apiVersion: kubecfg.dev/v1alpha1
        kind: AppInstance
        metadata:
          name: influxdb
          namespace: influxdb
        pause: false
        # ...
        ```

    2.  Apply the change to resume the `kubit` operator and scale InfluxDB
        Clustered components to the number of replicas defined for each in your
        `AppInstance` resource:

        ```bash
        kubectl apply --filename myinfluxdb.yml --namespace influxdb
        ```

    3.  Verify that InfluxDB Clustered pods start running again.

Your InfluxDB cluster is now restored to the recovery point.
When the Garbage Collector runs, it identifies what Parquet files are not
associated with the recovery point and [soft deletes](#soft-delete) them.

> ## Definitions

Reviewer: ```suggestion
```

> ### Soft delete
> 
> A soft delete refers to when, on compaction, the Compactor sets a `deleted_at`
> timestamp on the Parquet file entry in the Catalog. The Parquet file is no
> longer queryable, but remains intact in the object store. 
> 
> ### Hard delete
> 
> A hard delete refers to when a Parquet file is actually deleted from object
> storage and no longer exists.
> 

Reviewer: ```suggestion
```

> ### Recovery Point Objective (RPO)
> 
> The maximum amount of data (based on time) that can be lost after a disruptive event.
> RPO indicates how much time can pass between data snapshots before data is
> considered lost if a disaster occurs.
> 
> ### Recovery Time Objective (RTO)
> 
> The maximum amount of time that an InfluxDB cluster can be down after a failure.

Reviewer: ```suggestion
```
